platform-sdk/docs/proposals/001-tss-rosters-design-proposal/proposal.md

```diff
@@ -0,0 +1,447 @@
+There will continue to be components used in creating the Roster - such as the private EC (elliptic curve) key, the RSA
+private key and the signing X509Certificate certificate - that will be stored on disk.
+However, it is up to Services to manage the lifecycle of these files, and not the platform.
+
+There will not be any other separate artifacts stored elsewhere (e.g., directly on disk.).
+
+### Startup procedure (pseudo-code)
+
+The pseudocode for the startup procedure will look as follows:
+
+```java
+/**
+ * Start the platform.
+ * @param state an initial state. The caller either loads it from disk if the node has run before,
+ *              or constructs a new empty state and stores the genesis roster in there
+ *              as the current active roster.
+ */
+void startPlatform(final State state) {
+    if (the State has Candidate Roster) {
+        if (isSoftwareUpgrade) {
+            // This MUST be performed under `isSoftwareUpgrade` to ensure that
+            // the entire network is being restarted, and so every node adopts the Candidate Roster,
+            // and hence no ISSes happen.
+            // Modify the state and put Candidate Roster into Active Roster, effectively clearing the Candidate Roster.
+            makeCRtheAR();
+            // May make a record of the previous Active Roster if necessary (e.g. for PCES replay)
+        }
+    }
+
+    if (the State has no Active Roster) {
+        // This should never happen, but we have to check this because we're given a state as an argument.
+        throwFatalErrorAndShutdown();
+    }
+
+    // At this point the Active Roster in the state is what we'll be using as a roster.
+}
+```
+
+### Roster changes needed for Components
+
+#### Block Proof
+
+In a fully Dynamic Address Book (DAB) where, say, the roster for round 101 might be different from the roster for round
+102, there is a need to track what roster is used for each non-ancient round.
+We propose the introduction of a queue of roster hashes stored in the state, where the index of a roster hash determines
+what round the roster is active for as shown.
+![](TSS%20Roster%20Lifecycle-Roster%20and%20Rounds.drawio.svg)
+The size of this queue would be ancient round window + roster offset.
+The roster offset refers to the number of rounds in advance that the network determines and declares the active roster
+for a specific future round.
+This acts as a proactive mechanism that allows the network to prepare for upcoming changes in the roster and ensure a
+valid transition to a future roster.
+When each round is handled, we will pop off the active roster hash for the round that just became ancient, and add a new
+active roster hash for a new future round.
+
+Before full DAB (DAB Phase 2.5), we propose to store the current and previous active rosters in this newly introduced
+queue of hashes.
+In the DAB Phase 2.5 implementation, this queue will contain only 2 hashes, with the first hash representing
```

Comment:
The term `DAB Phase 2.5` is not referencing a formal deliverable.  It's an informal term I've used to refer to the changes that we're incorporating.  It may confuse external reviewers (PM, Services, DevOps, etc) that are not onboarded with the meaning.   It might be better to simply remove the term and say `In this proposal implementation ...`

Reply:
updated, thanks!